paasta_tools/check_autoscaler_max_instances.py

```diff
@@ -0,0 +1,159 @@
+                # TODO: should we send status=0 here, in case someone disables autoscaling for their service / changes
+                # to bespoke autoscaler?
+                continue
+
+            if not job_config.get_docker_image():
+                # skip services that haven't been marked for deployment yet.
+                continue
+
+            autoscaling_status = await pik.autoscaling_status(
+                kube_client=kube_client,
+                job_config=job_config,
+                namespace=job_config.get_namespace(),
+            )
+            if autoscaling_status["min_instances"] == -1:
+                log.warning(
+                    f"HPA {job_config.get_sanitised_deployment_name()} not found."
+                )
+                continue
+
+            if (
+                autoscaling_status["min_instances"]
+                == autoscaling_status["max_instances"]
+            ):
+                status = pysensu_yelp.Status.OK
+                output = f"Not checking {service}.{instance} as min_instances == max_instances == {autoscaling_status['max_instances']}. This is probably a canary instance."
+            elif (
+                autoscaling_status["desired_replicas"]
+                >= autoscaling_status["max_instances"]
+            ):
+                threshold = job_config.get_autoscaling_max_instances_alert_threshold()
+                setpoint = job_config.get_autoscaling_params()["setpoint"]
+                metric_threshold_target_ratio = threshold / setpoint
+
+                status = pysensu_yelp.Status.UNKNOWN
+                output = "how are there no metrics for this thing?"
+                for metric in autoscaling_status["metrics"]:
+                    current_value = suffixed_number_value(metric["current_value"])
+                    target_value = suffixed_number_value(metric["target_value"])
+
+                    if current_value / target_value > metric_threshold_target_ratio:
+                        status = pysensu_yelp.Status.CRITICAL
+                        output = f"{service}.{instance}: Ratio of current value to target value ({current_value} / {target_value}) is greater than the ratio of max_instances_alert_threshold to setpoint ({threshold} / {setpoint})"
+                    else:
+                        status = pysensu_yelp.Status.OK
+                        output = f"{service}.{instance}: Ratio of current value to target value ({current_value} / {target_value}) is below the ratio of max_instances_alert_threshold to setpoint ({threshold} / {setpoint})"
+            else:
+                status = pysensu_yelp.Status.OK
+                output = f"{service}.{instance} is below max_instances."
+
+            monitoring_overrides = job_config.get_monitoring()
+            monitoring_overrides.update(
+                {
+                    "page": False,  # TODO: remove this line once this alert has been deployed for a little while.
```

Comment:
might be worth ticketing this (and adding a note in the ticket that we'll need to send an email letting folks know that there's a new paging alert being added)